src/stdlib_linalg.fypp

```diff
@@ -82,20 +82,28 @@ module stdlib_linalg
 
 contains
 
-    function eye(n) result(res)
-      !! version: experimental
-      !!
-      !! Constructs the identity matrix
-      !! ([Specification](../page/specs/stdlib_linalg.html#description_1))
-      integer, intent(in) :: n
-      integer(int8) :: res(n, n)
-      integer :: i
-      res = 0
-      do i = 1, n
-         res(i, i) = 1
-      end do
-    end function eye
+    !> Version: experimental
+    !>
+    !> Constructs the identity matrix.
+    !> ([Specification](../page/specs/stdlib_linalg.html#eye-construct-the-identity-matrix))
+    pure function eye(dim1, dim2) result(result)
+
+        integer, intent(in) :: dim1
+        integer, intent(in), optional :: dim2
+        integer, allocatable :: result(:, :)
+
+        integer :: dim2_
+        integer :: i
 
+        dim2_ = merge(dim2, dim1, present(dim2))
```

Comment:
Please use `optval` instead of `merge` here to avoid segmentation faults with compilers that don't extend `merge` with short circuiting behaviour.